### PR TITLE
ORC-475: ORC reader should lazily get filesystem

### PR DESCRIPTION
**Pull request by prasanthj**

### CODE REVIEW

File: java/core/src/java/org/apache/orc/impl/ReaderImpl.java [hunk 1]
```diff
@@ -365,6 +364,14 @@ public ReaderImpl(Path path, OrcFile.ReaderOptions options) throws IOException {
     this.schema = OrcUtils.convertTypeFromProtobuf(this.types, 0);
   }
 
+  protected FileSystem getFileSystem() throws IOException {
```
Comment (t3rmin4t0r) [inline java/core/src/java/org/apache/orc/impl/ReaderImpl.java]: Two calls to getFileSystem() will do this 2x - have this init this.fileSystem inline.

Comment (t3rmin4t0r) [reply java/core/src/java/org/apache/orc/impl/ReaderImpl.java]: Also use the options.filesystem if it provided

File: java/core/src/java/org/apache/orc/impl/ReaderImpl.java [hunk 2]
```diff
@@ -341,7 +338,9 @@ public ReaderImpl(Path path, OrcFile.ReaderOptions options) throws IOException {
     } else {
       OrcTail orcTail = options.getOrcTail();
       if (orcTail == null) {
-        tail = extractFileTail(fs, path, options.getMaxLength());
+        this.fileSystem = getFileSystem();
+        options.filesystem(fileSystem);
```
Comment (t3rmin4t0r) [inline java/core/src/java/org/apache/orc/impl/ReaderImpl.java]: options fs object got lost - it's not clear that was null before.

File: java/core/src/java/org/apache/orc/impl/ReaderImpl.java (hunk 1 above)
Comment (prasanthj) [reply java/core/src/java/org/apache/orc/impl/ReaderImpl.java]: fixed it.

File: java/core/src/java/org/apache/orc/impl/ReaderImpl.java (hunk 2 above)
Comment (prasanthj) [reply java/core/src/java/org/apache/orc/impl/ReaderImpl.java]: fixed it.

**Pull request closed without merging**